File: CCC_2022/CCC_2022_Problem_J5_Square_Pool.java
import java.util.*;
public class CCC_2022_Problem_J5_Square_Pool {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt(), t = sc.nextInt(), ans = 0;
		if(t == 0) {
			System.out.println(n);
			sc.close();
			return;
		}
		List<pair> tree = new ArrayList<>();
		tree.add(new pair(0, 0));
		for(int i=0; i<t; i++) {
			int a = sc.nextInt(), b = sc.nextInt();
			tree.add(new pair(a, b));
		}
		for(pair pi: tree){
	        int top = pi.first;
	        for(pair pj: tree){
	            int left = pj.second;
	            int sz = Math.min(n-top, n-left);
	            for(pair pk: tree){
	                if(pk.first > top && pk.second > left){
	                    sz = Math.min(sz, Math.max(pk.first-top-1, pk.second-left-1));
	                }
	            }
	            ans = Math.max(ans, sz);
	        }
	    }
		System.out.println(ans);
		sc.close();
	}
	static class pair {
        public int first, second;
        public pair(int first, int second) {this.first = first; this.second = second;}
    }
}

File: CCC_2022/CCC_2022_Problem_J5_Square_Pool.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
    int n, t, ans = 0;
    cin >> n >> t;
    if(t == 0){
        cout << n << endl;
        return 0;
    }
    vector<pair<int, int>> tree;
    tree.push_back({0, 0});
    for(int i=0, a, b; i<t; i++){
        cin >> a >> b;
        tree.push_back({a, b});
    }
    for(pair<int, int> pi: tree){
        int top = pi.first;
        for(pair<int, int> pj: tree){
            int left = pj.second;
            int sz = min(n-top, n-left);
            for(pair<int, int> pk: tree){
                if(pk.first > top && pk.second > left){
                    sz = min(sz, max(pk.first-top-1, pk.second-left-1));
                }
            }
            ans = max(ans, sz);
        }
    }
    cout << ans << endl;
    return 0;
}
